### PR TITLE
chore(manifest): add sidecar image fields

### PR DESCRIPTION
**Pull request by KollaAdithya**
This PR adds the manifest fields for sidecar. 
Related to #4254 
Here is the proposed manifest Ux:
```yaml
sidecars:
  nginx:
    image: 'url'
```
  or
```yaml
sidecars:
  nginx:
    image:
      location: 'url'
```
or
```yaml
sidecars:
  nginx:
    image:
      build: 'path/to/Dockerfile'
```
or
```yaml
sidecars:
  nginx:
    image:
      build:
        dockerfile: path/to/Dockerfile
        context: context/dir
        target: build-stage
        cache_from:
          - image:tag
        args:
          key: value
```

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the Apache 2.0 License.

### CODE REVIEW

File: internal/pkg/manifest/workload_ecs.go [hunk 1]
```diff
@@ -299,19 +299,56 @@ func (lc *Logging) GetEnableMetadata() *string {
...
+
+// UnmarshalYAML overrides the default YAML unmarshaling logic for the Image
+// struct, allowing it to perform more complex unmarshaling behavior.
+// This method implements the yaml.Unmarshaler (v3) interface.
+func (s *SidecarImageConfig) UnmarshalYAML(value *yaml.Node) error {
+	type scImageConfig SidecarImageConfig
+	if err := value.Decode((*scImageConfig)(s)); err != nil {
+		return err
+	}
+	if !s.Build.IsZero() && s.Location != nil {
```
Comment (paragbhingre) [inline internal/pkg/manifest/workload_ecs.go]: What I think is this validation should be inside manifest/validation.go and not here in the `UnmarshalYAML`.

File: internal/pkg/manifest/workload_ecs_test.go [hunk 2]
```diff
@@ -478,3 +509,39 @@ func TestLogging_GetEnableMetadata(t *testing.T) {
...
 	}
 }
+
+func Test_SidecarImageURI(t *testing.T) {
+	testCases := map[string]struct {
+		in     SidecarConfig
+		wanted *string
+	}{
+		"empty SidecarConfig": {},
+		"should return URI is provided directly through `image` ": {
```
Comment (paragbhingre) [inline internal/pkg/manifest/workload_ecs_test.go]:
```suggestion
		"should return URI if provided directly through `image` ": {
```

File: internal/pkg/manifest/validate.go [hunk 3]
```diff
@@ -2042,3 +2045,8 @@ func contains(name string, names []string) bool {
 	}
 	return false
 }
+
+// validate is a no-op for SidecarImageConfig.
+func (s SidecarImageConfig) validate() error {
```
Comment (paragbhingre) [inline internal/pkg/manifest/validate.go]: This is the place where you should be testing what we are testing inside `UnmarshalYAML`

File: internal/pkg/manifest/validate.go [hunk 4]
```diff
@@ -1326,6 +1326,9 @@ func (s SidecarConfig) validate() error {
 	if err := s.DependsOn.validate(); err != nil {
 		return fmt.Errorf(`validate "depends_on": %w`, err)
 	}
+	if err := s.Image.Advanced.validate(); err != nil {
+		return fmt.Errorf(`validate "sidecar_imageconfig: %w`, err)
```
Comment (paragbhingre) [inline internal/pkg/manifest/validate.go]: We always showcase the exact manifest field to the customer that requires fixing. 

```suggestion
		return fmt.Errorf(`validate "build: %w`, err)
```

File: internal/pkg/manifest/backend_svc_test.go [hunk 5]
```diff
@@ -1098,8 +1108,10 @@ func TestBackendService_ExposedPorts(t *testing.T) {
 					},
 					Sidecars: map[string]*SidecarConfig{
 						"xray": {
-							//Port:       aws.String("2000"),
-							Image:      aws.String("123456789012.dkr.ecr.us-east-2.amazonaws.com/xray-daemon"),
+							//Port: aws.String("2000"),
+							Image: Union[*string, SidecarImageConfig]{
+								Basic: aws.String("123456789012.dkr.ecr.us-east-2.amazonaws.com/xray-daemon"),
```
Comment (paragbhingre) [inline internal/pkg/manifest/backend_svc_test.go]: For one of the test we can also try putting this URI inside `image.location` instead of `image`

**Review by paragbhingre — COMMENTED** [review]
I just have a few nits and a couple of requests. Overall, it looks good to me.

File: internal/pkg/manifest/backend_svc_test.go (hunk 5 above)
Comment (efekarakus) [reply internal/pkg/manifest/backend_svc_test.go]: Yeah that'd be awesome! Perhaps not in `_ExposedPorts` tests but within the `transform_test`?

File: internal/pkg/manifest/workload_ecs.go [hunk 6]
```diff
@@ -299,19 +299,56 @@ func (lc *Logging) GetEnableMetadata() *string {
...
+			mustExist:   true,
+		}
+	}
+	return nil
+}
+
+// SidecarImageURI returns the URI of the image.
+func (s *SidecarConfig) SidecarImageURI() *string {
+	// Prefer to use the "Image" string in SidecarConfig. Otherwise,
+	// "Location" in SidecarImageConfig. If no image URI is specified, return nil.
```
Comment (efekarakus) [inline internal/pkg/manifest/workload_ecs.go]: This sort of info would be useful to the client! Therefore, it'd be neat to move it higher up into the method doc comments

File: internal/pkg/manifest/workload_ecs.go [hunk 7]
```diff
@@ -299,19 +299,56 @@ func (lc *Logging) GetEnableMetadata() *string {
...
+			firstField:  "build",
+			secondField: "location",
+			mustExist:   true,
+		}
+	}
+	return nil
+}
+
+// SidecarImageURI returns the URI of the image.
+func (s *SidecarConfig) SidecarImageURI() *string {
```
Comment (efekarakus) [inline internal/pkg/manifest/workload_ecs.go]: What do you think of this method signature:
```
// ImageURI returns the location of the image if one is set. 
// If the image needs to be build, return "" and false.
func (cfg *SidecareConfig) ImageURI() (string, bool)
```
It's like looking up a value from a hashmap `v, ok := map[key]`

File: internal/pkg/manifest/svc_test.go [hunk 8]
```diff
@@ -147,7 +147,7 @@ environments:
 						Sidecars: map[string]*SidecarConfig{
 							"xray": {
 								Port:       aws.String("2000/udp"),
-								Image:      aws.String("123456789012.dkr.ecr.us-east-2.amazonaws.com/xray-daemon"),
+								Image:      BasicToUnion[*string, SidecarImageConfig](aws.String("123456789012.dkr.ecr.us-east-2.amazonaws.com/xray-daemon")),
```
Comment (efekarakus) [inline internal/pkg/manifest/svc_test.go:160]: It'd be amazing if we can add a test case here where the `inContent` has the advanced configuration for the sidecar and we ensure that we're unmarshaling properly into `SidecarConfig`

**Review by efekarakus — COMMENTED** [review]
Looks great!

File: internal/pkg/manifest/workload_ecs.go [hunk 9]
```diff
@@ -299,19 +299,56 @@ func (lc *Logging) GetEnableMetadata() *string {
...
-	Essential     *bool                `yaml:"essential"`
-	CredsParam    *string              `yaml:"credentialsParameter"`
-	Variables     map[string]Variable  `yaml:"variables"`
-	Secrets       map[string]Secret    `yaml:"secrets"`
-	MountPoints   []SidecarMountPoint  `yaml:"mount_points"`
-	DockerLabels  map[string]string    `yaml:"labels"`
-	DependsOn     DependsOn            `yaml:"depends_on"`
-	HealthCheck   ContainerHealthCheck `yaml:"healthcheck"`
+	Port          *string                            `yaml:"port"`
+	Image         Union[*string, SidecarImageConfig] `yaml:"image"`
```
Comment (huanjani) [inline internal/pkg/manifest/workload_ecs.go:303]: nice!

File: internal/pkg/manifest/backend_svc_test.go (hunk 5 above)
Comment (KollaAdithya) [reply internal/pkg/manifest/backend_svc_test.go]: Do we still need custom Transformer for mergo in `transform.go` and `transform_test` 🤔

File: internal/pkg/manifest/backend_svc_test.go [hunk 10]
```diff
@@ -1098,8 +1108,12 @@ func TestBackendService_ExposedPorts(t *testing.T) {
 					},
 					Sidecars: map[string]*SidecarConfig{
 						"xray": {
-							//Port:       aws.String("2000"),
-							Image:      aws.String("123456789012.dkr.ecr.us-east-2.amazonaws.com/xray-daemon"),
+							//Port: aws.String("2000"),
```
Comment (huanjani) [inline internal/pkg/manifest/backend_svc_test.go]: Should this still be commented out? 🤔

**Review by huanjani — APPROVED** [review]
Just one tiny question; otherwise, LGTM!

File: internal/pkg/manifest/backend_svc_test.go (hunk 10 above)
Comment (KollaAdithya) [reply internal/pkg/manifest/backend_svc_test.go]: removed the commented line

File: internal/pkg/manifest/workload_ecs.go [hunk 11]
```diff
@@ -299,19 +299,37 @@ func (lc *Logging) GetEnableMetadata() *string {
...
 
+// SidecarImageConfig represents the docker build arguments and location of the existing image.
+type SidecarImageConfig struct {
+	Build    Union[*string, DockerBuildArgs] // `yaml:"build"`
+	Location *string                         // `yaml:"location"`
+}
+
+// ImageURI returns the location of the image if one is set.
+// If the image needs to be build, return "" and false.
+func (cfg *SidecarConfig) ImageURI() (string, bool) {
```
Comment (iamhopaul123) [inline internal/pkg/manifest/workload_ecs.go:323]: Are we using the `bool` from output anywhere 🤔

Comment (KollaAdithya) [reply internal/pkg/manifest/workload_ecs.go:323]: No we are not using `bool` anywhere. As per Efe's [recommendation ](https://github.com/aws/copilot-cli/pull/4430#discussion_r1091369434) I changed to this method signature. 
I will change it back to 
`func (cfg *SidecarConfig) ImageURI() *string` if you insist.

**Review by iamhopaul123 — APPROVED** [review]

File: internal/pkg/manifest/workload_ecs.go [hunk 12]
```diff
@@ -299,19 +299,37 @@ func (lc *Logging) GetEnableMetadata() *string {
...
+	Secrets       map[string]Secret                  `yaml:"secrets"`
+	MountPoints   []SidecarMountPoint                `yaml:"mount_points"`
+	DockerLabels  map[string]string                  `yaml:"labels"`
+	DependsOn     DependsOn                          `yaml:"depends_on"`
+	HealthCheck   ContainerHealthCheck               `yaml:"healthcheck"`
 	ImageOverride `yaml:",inline"`
 }
 
+// SidecarImageConfig represents the docker build arguments and location of the existing image.
+type SidecarImageConfig struct {
```
Comment (Lou1415926) [inline internal/pkg/manifest/workload_ecs.go:316]: sorry I am late to the party 😂 

This is not a request to change the implementation but just an idea. What do you think of doing this:
```go
type ImageLocationOrBuild struct {
	Build    Union[*string, DockerBuildArgs] `yaml:"build"`
	Location *string                                         `yaml:"location"`
}
type Image struct {
      ImageLocationOrBuild `yaml:",inline"`
}

type SidecarConfig struct {
     Image         Union[*string, ImageLocationOrBuild] `yaml:"image"`
}
```

This struct here is named `SidecarImageConfig` but there is actually nothing specific about `Sidecar` in it. It looks like a general purpose struct that can be used for main container as well.

**Pull request merged**